### PR TITLE
Restart required

### LINKED ISSUE #24: Deduce target based on var files change (restart required)
When some var files change, deduce target nodes on the DAG to reach.

Need:
- [x] #102 
- [x] #107 
- [x] #108

### PR DESCRIPTION
**Pull request by rpignolet**
Fix #24 

Add restart required feature

Add a new CLI command `restart-required` which configure and restart needed services/components if a service/component DAG ancestors configuration is modified.

### CODE REVIEW

File: tdp/cli/commands/restart_required.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        operation_runner = OperationRunner(dag, ansible_executor, service_managers)
+
+        deployment = operation_runner.run_nodes(
+            sources=list(components_modified),
+            restart=True,
+        )
+
+        if not dry:
+            session.add(deployment)
+            session.commit()
```
Comment (gboutry) [inline tdp/cli/commands/restart_required.py:97]: In the deploy command, even if it's dry, it's inserted in the database. Is there a reason this behavior is different in this command ?

File: tdp/cli/commands/restart_required.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 TOSIT.IO
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from pathlib import Path
+
+import click
+
+from tdp.cli.commands.service_versions import get_latest_success_service_version
```
[the remark this answers is not in the document]
Comment (gboutry) [reply tdp/cli/commands/restart_required.py:9]: Moreover, I think `get_latest_success_service_version` should become `get_latest_success_service_version_query`, not take the session as parameter and return a selectable.

File: tdp/cli/commands/restart_required.py (hunk 1 above)
Comment (gboutry) [reply tdp/cli/commands/restart_required.py:97]: ref #140

File: tdp/cli/commands/restart_required.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/cli/commands/restart_required.py:9]: Ok but in the actual implementation the `session` is used for `session.query()` how I replace this ?

File: tdp/cli/commands/restart_required.py (hunk 1 above)
Comment (rpignolet) [reply tdp/cli/commands/restart_required.py:97]: This is not the behavior that we want ? In `dry` mode it should not write in the database or I did not understand the #140.

File: tdp/cli/commands/restart_required.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/cli/commands/restart_required.py]: I move the function inside a `queries` file but I don't know how to transform `session.query` to `select`.

[the remark this answers is not in the document]
Comment (gboutry) [reply tdp/cli/commands/restart_required.py]: You change the part `session.query(...).all()` to `select(...)` (without the all).

Then, when you use it:
`session.execute(the selectable object).all()`

File: tdp/cli/commands/restart_required.py (hunk 1 above)
Comment (gboutry) [reply tdp/cli/commands/restart_required.py:97]: Well the issue #40 questions it, but we didn't statuate on this yet. You're introducing new behavior before anything has been decided.

File: tdp/cli/commands/restart_required.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpignolet) [reply tdp/cli/commands/restart_required.py]: Ok done.

File: tdp/cli/commands/restart_required.py (hunk 1 above)
Comment (rpignolet) [reply tdp/cli/commands/restart_required.py]: I deleted the `if`.

**Review by gboutry — APPROVED** [review]

File: tdp/cli/commands/restart_required.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                raise RuntimeError(
+                    f"Service '{service}' is deployed but the repository is missing."
+                )
+            for component_modified in service_managers[service].components_modified(
+                version
+            ):
+                components_modified.add(component_modified.name)
+
+        if len(components_modified) == 0:
+            click.echo("Nothing need to be restarted")
```
Comment (gboutry) [inline tdp/cli/commands/restart_required.py:87]:
```suggestion
            click.echo("Nothing needs to be restarted")
```

Comment (rpignolet) [reply tdp/cli/commands/restart_required.py]: done

**Pull request merged**